File: polynomial_regression.py
# Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv("Position_Salaries.csv")
X = dataset.iloc[:, 1:2].values # 2 is used only for x as matrix not vector
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
# No need to split the dataset since we have few data

# Feature Scaling
# note : regression fit method deals with feature scaling automatically

# 2 models are produced to compare them

# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y) # not X_train or y_train

# Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4) # auto add ones col at the left of the matrix
X_poly = poly_reg.fit_transform(X)
poly_reg.fit(X_poly, y)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualizing the Linear Regression results
plt.scatter(X, y, color = "red")
plt.plot(X, lin_reg.predict(X), color = "blue")
plt.title("Truth or Bluff (Linear Regression)")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Visualizing the Polynomial Regression results
plt.scatter(X, y, color = "red")
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color = "blue")
plt.title("Truth or Bluff (Polynomial Regression)")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Visualizing the Polynomial Regression results (for higher resolution and smoother curve)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = "red")
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = "blue")
plt.title("Truth or Bluff (Polynomial Regression)")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Predicting a new result with Linear Regression
lin_reg.predict([[7.5]])

# Predicting a new result with Polynomial Regression
lin_reg_2.predict(poly_reg.fit_transform([[7.5]]))

File: polynomial_regression.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv("Position_Salaries.csv")
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# No need to split the dataset since we have few data

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Linear Regression to the dataset
lin_reg = lm(formula = Salary ~ .,
             data = dataset)
summary(lin_reg)

# Fitting Polynomial Regression to the dataset

# ---------- this part is iterative ---------- #
# you always check the summary and the graph to stop adding polynomial features
# add polynomial features
dataset$Level2 = dataset$Level ^ 2
dataset$Level3 = dataset$Level ^ 3
dataset$Level4 = dataset$Level ^ 4
# you can continue with levels of 5 and 6
# dataset$Level5 = dataset$Level ^ 5
# dataset$Level6 = dataset$Level ^ 6

poly_reg = lm(formula = Salary ~ .,
              data = dataset)
summary(poly_reg)

# Visualizing the Linear Regression results
install.packages('ggplot2')
library(ggplot2)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = "blue") +
  ggtitle(label = "Truth or Bluff (Linear Regression)") +
  xlab("Level") +
  ylab("Salary") +
  theme(plot.title = element_text(hjust = 0.5, vjust = 0.5))

# Visualizing the Polynomial Regression results
install.packages('ggplot2')
library(ggplot2)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = "blue") +
  ggtitle("Truth or Bluff (Polynomial Regression)") +
  xlab("Level") +
  ylab("Salary") +
  theme(plot.title = element_text(hjust = 0.5, vjust = 0.5))

# Visualizing the Regression Model results (for higher resolution and smoother curve)
install.packages('ggplot2')
library(ggplot2)

x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = x_grid, y = predict(
    poly_reg,
    newdata = data.frame(
      Level = x_grid,
      Level2 = x_grid ^ 2,
      Level3 = x_grid ^ 3,
      Level4 = x_grid ^ 4
      #,Level5 = x_grid ^ 5,
      #Level6 = x_grid ^ 6
      ))),
  colour = "blue") +
  ggtitle("Truth or Bluff (Polynomial Regression)") +
  xlab("Level") +
  ylab("Salary")

# ----------- iterative part end --------------- #

# Predicting a new result with Linear Regression
y_pred = predict(lin_reg, data.frame(Level = 7.5))

# Predicting a new result with Polynomial Regression
y_pred_poly = predict(poly_reg,
                 data.frame(
                   Level = 7.5,
                   Level2 = 7.5 ^ 2,
                   Level3 = 7.5 ^ 3,
                   Level4 = 7.5 ^ 4
                   #,Level5 = x_grid ^ 5,
      			   #Level6 = x_grid ^ 6
                 ))
